### PR TITLE
New: Rule `https-only`

### PR DESCRIPTION
**Pull request by sarvaje**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

### CODE REVIEW

File: packages/rule-https-only/package.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    "failFast": false,
+    "files": [
+      "dist/tests/**/*.js"
+    ],
+    "timeout": "1m"
+  },
+  "description": "sonarwhal rule that checks if your site is using https and if it has Mixed Content.",
+  "devDependencies": {
+    "@types/debug": "0.0.30",
+    "@types/node": "8.0.14",
```
Comment (alrra) [inline packages/rule-https-only/package.json]: Do you need this? We removed it from everywhere else as it creates a lot of problems.

File: packages/rule-https-only/package.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    "lint:md": "markdownlint --ignore CHANGELOG.md *.md",
+    "sonarwhal": "node node_modules/sonarwhal/dist/src/bin/sonarwhal.js",
+    "test": "npm run lint && npm run build && nyc ava",
+    "init": "npm install && npm run build",
+    "watch": "npm run build && npm-run-all --parallel -c watch:*",
+    "watch:assets": "npm run build:assets -- -w --no-initial",
+    "watch:test": "ava --watch",
+    "watch:ts": "npm run build:ts -- --watch"
+  },
+  "peerDependencies": {
```
Comment (alrra) [inline packages/rule-https-only/package.json]: Can you sort things alphabetically? Thanks!

File: packages/rule-https-only/src/rule.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+/**
+ * @fileoverview sonarwhal rule that checks if your site is using https and if it has Mixed Content.
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: > HTTPS

File: packages/rule-https-only/src/rule.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    private targetResource: string = '';
+
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.security,
+            description: `sonarwhal rule that checks if your site is using https and if it has Mixed Content.`
+        },
+        id: 'https-only',
+        schema: [
+            /*
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: Remove this.

File: packages/rule-https-only/src/rule.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+export default class HttpsOnlyRule implements IRule {
+    private targetHTTPs: boolean = false;
+    private targetResource: string = '';
+
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.security,
+            description: `sonarwhal rule that checks if your site is using https and if it has Mixed Content.`
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: > HTTPS

File: packages/rule-https-only/src/rule.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+        const validateTarget = async (event: Event): Promise<boolean> => {
+            const { resource } = event;
+
+            this.targetResource = resource;
+
+            if (!isHTTPS(resource)) {
+                debug('HTTPS no detected');
+
+                await context.report(resource, null, 'The site should be https');
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: > Should be served over HTTPS.

Can you use `HTTPS` everywhere else? Thanks!

File: packages/rule-https-only/package.json [hunk 7]
```diff
@@ -0,0 +1,70 @@
+{
+  "ava": {
+    "failFast": false,
+    "files": [
+      "dist/tests/**/*.js"
+    ],
+    "timeout": "1m"
+  },
+  "description": "sonarwhal rule that checks if your site is using https and if it has Mixed Content.",
```
Comment (alrra) [inline packages/rule-https-only/package.json]: > HTTPS 

> mixed content

File: packages/rule-https-only/src/rule.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+                return false;
+            }
+
+            debug('HTTPS detected');
+
+            this.targetHTTPs = true;
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: Maybe use something more descriptive, like: `targetIsServedOverHTTPS` ?

File: packages/rule-https-only/tests/tests.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
+import { readFileSync } from 'fs';
+
+import { generateHTMLPage } from 'sonarwhal/dist/tests/helpers/misc';
+import { RuleTest } from 'sonarwhal/dist/tests/helpers/rule-test-type';
+import * as ruleRunner from 'sonarwhal/dist/tests/helpers/rule-runner';
+
+const ruleName = 'https-only';
```
Comment (alrra) [inline packages/rule-https-only/tests/tests.ts]: Use `getRuleName`, don't hardcode the value.

File: packages/rule-https-only/src/rule.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                return;
+            }
+
+            if (!this.targetHTTPs) {
+                return;
+            }
+
+            const { resource } = event;
+
+            if (resource === this.targetResource) {
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: Why is this check necessary?

File: packages/rule-https-only/src/rule.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                return;
+            }
+
+            if (!isHTTPS(resource)) {
+                await context.report(resource, null, 'Requested insecure content');
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: Why not just used `Should be served over HTTPS`?

File: packages/rule-https-only/README.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+It's good to put some edge cases in here.
+
+## Can the rule be configured?
+
+If this rule allows some configuration, please put the format and
+options for the user.
+
+## Further Reading
+
+What can the user read to know more about this subject?
```
Comment (alrra) [inline packages/rule-https-only/README.md]: When you write the docs, add this https://doesmysiteneedhttps.com/ (and also feel free to get some inspiration from it).

File: packages/rule-https-only/package.json (hunk 1 above)
Comment (sarvaje) [reply packages/rule-https-only/package.json]: We have it in the root one, maybe it is not necessary here.

File: packages/rule-https-only/src/rule.ts (hunk 10 above)
Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: yep, maybe it is not necessary

File: packages/rule-https-only/src/rule.ts (hunk 11 above)
Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: I like it (I'm not very good for messages heheh)

File: packages/rule-https-only/README.md (hunk 12 above)
Comment (sarvaje) [reply packages/rule-https-only/README.md]: ok!

File: packages/rule-https-only/README.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
+# https-only (`https-only`)
+
+`https-only` check if your site is using HTTPS and warns against having
```
Comment (molant) [inline packages/rule-https-only/README.md]: > checks

File: packages/rule-https-only/README.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+mixed content.
+
+## Why is this important?
+
+HTTPS is important to guarantees content integrity. Even when your site
```
Comment (molant) [inline packages/rule-https-only/README.md]: > guarantee

File: packages/rule-https-only/README.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+doesn't have sensitive information, an attacker can change the content
+or inject scripts.
```
Comment (molant) [inline packages/rule-https-only/README.md]: > an attacker could change the content or inject malicious scripts (like a [crypto miner](https://scotthelme.co.uk/protect-site-from-cryptojacking-csp-sri/) to use your users CPU power).

File: packages/rule-https-only/README.md [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    ...
+}
+```
+
+## What does the rule check?
+
+This rule checks two things:
+
+* The url is served using HTTPS
```
Comment (molant) [inline packages/rule-https-only/README.md]: > The main target is served using HTTPS
> If the main target is an HTML file, all its resources should be on HTTPS too

File: packages/rule-https-only/README.md [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+* If so, all the content served is HTTPS too.
+
+### Examples that **trigger** the rule
+
+If your site is not served as HTTPS.
```
Comment (molant) [inline packages/rule-https-only/README.md]: > using HTTPS

File: packages/rule-https-only/README.md [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+```
+sonarwhal http://example.com
+```
+
+If your site is served as HTTPS, but your assets are HTTP.
```
Comment (molant) [inline packages/rule-https-only/README.md]: > using HTTPS, but one or more resources use HTTP.

File: packages/rule-https-only/README.md [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+```html
+<body>
+    <img src="http://assets.com/image.png" />
+    <script src="http://assets.com/script.js"></script>
+</body>
+```
+
+### Examples that **pass** the rule
+
+Your site is served as HTTPS and your assets too.
```
Comment (molant) [inline packages/rule-https-only/README.md]: > using HTTPS and its resources too.

File: packages/rule-https-only/README.md [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+```html
+<body>
+    <img src="https://assets.com/image.png" />
+    <script src="https://assets.com/script.js"></script>
+</body>
+```
+
+## Can the rule be configured?
+
+This rule cannot be configured.
```
Comment (molant) [inline packages/rule-https-only/README.md]: I'd remove this section. Don't think we have it for rules that can't be configured.

File: packages/rule-https-only/package.json [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+  "ava": {
+    "failFast": false,
+    "files": [
+      "dist/tests/**/*.js"
+    ],
+    "timeout": "1m"
+  },
+  "description": "sonarwhal rule that checks if your site is using HTTPS and if it has mixed content.",
+  "devDependencies": {
+    "@types/debug": "0.0.30",
```
Comment (molant) [inline packages/rule-https-only/package.json]: Do we need the `debug` types? Aren't they pulled from the main?

File: packages/rule-https-only/src/rule.ts [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+            }
+
+            const { resource } = event;
+
+            if (!isHTTPS(resource)) {
+                await context.report(resource, null, 'Should be served over HTTPS');
+            }
+        };
+
+        context.on('fetch::start::target', validateTarget);
```
Comment (molant) [inline packages/rule-https-only/src/rule.ts]: I think we should subscribe to the `fetch::end::target` because there could be a redirect from HTTP to HTTPS and I don't think we should penalize that. If there are redirects for resources we should fail those.

Comment (molant) [reply packages/rule-https-only/src/rule.ts]: Also I would subscribe to all `fetch::end`. There could be a weird redirect case like HTTPS -->HTTP that is not detected here so we should check that all the `hops` are on HTTPS.

File: packages/rule-https-only/tests/tests-https.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+const generateResponse = (content: Buffer, type: string): Object => {
+    return {
+        content,
+        headers: { 'Content-Type': type }
+    };
+};
+
+const testsHTTPS: Array<RuleTest> = [
+    {
+        name: 'This test should pass',
```
Comment (molant) [inline packages/rule-https-only/tests/tests-https.ts]: > HTML page with no resources over HTTPS passes

File: packages/sonarwhal/src/lib/connectors/debugging-protocol-common/debugging-protocol-connector.ts [hunk 24]
```diff
@@ -673,9 +673,11 @@ export class Connector implements IConnector {
 
         try {
             debug(`resource ${href} to be fetched`);
-            await this._server.emitAsync('fetch::start', { resource: href });
+            const fullFaviconUrl = this._finalHref + href.substr(1);
```
Comment (molant) [inline packages/sonarwhal/src/lib/connectors/debugging-protocol-common/debugging-protocol-connector.ts]: The favicon part doesn't seem to be part of this PR?

Comment (sarvaje) [reply packages/sonarwhal/src/lib/connectors/debugging-protocol-common/debugging-protocol-connector.ts]: It is part, this PR has some changes in sonarwhal too (in a different commit)

File: packages/rule-https-only/src/rule.ts (hunk 22 above)
Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: There is no `fetch::end::target` event, but probably it would be nice to have it.

> Also I would subscribe to all fetch::end. There could be a weird redirect case like HTTPS -->HTTP that is not detected here so we should check that all the hops are on HTTPS.

I think that starting a fetch to `http` can be also a problem, I think that an http petition is vulnerable even if it redirect to a `https`. But I'm not sure...

File: packages/sonarwhal/src/lib/connectors/debugging-protocol-common/debugging-protocol-connector.ts (hunk 24 above)
Comment (molant) [reply packages/sonarwhal/src/lib/connectors/debugging-protocol-common/debugging-protocol-connector.ts]: OK!

File: packages/rule-https-only/package.json (hunk 21 above)
Comment (sarvaje) [reply packages/rule-https-only/package.json]: I will take a look

File: packages/rule-https-only/src/rule.ts (hunk 22 above)
Comment (molant) [reply packages/rule-https-only/src/rule.ts]: Yes, but we already have some rules that touch on this like [STS](https://sonarwhal.com/docs/user-guide/rules/rule-strict-transport-security/) so I think we should be ok with that. Otherwise you will need to know reuse a lot of the code in there, verify if the site is in the list, etc.

File: packages/rule-https-only/src/rule.ts [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        context.on('fetch::start', validate);
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: Shouldn't the manifest be included?

Comment (molant) [reply packages/rule-https-only/src/rule.ts]: Also don't we need to subscribe using fetch::start::* ?

File: packages/rule-https-only/src/rule.ts (hunk 22 above)
Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: Then, I need to add the event `fetch::end::target` and change the events to `fetch::end::*`, right?

File: packages/rule-https-only/src/rule.ts (hunk 25 above)
Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: we only have `fetch::start`, `fetch::start::target` and `fetch::start::manifest`, `fetch::start::*` doesn't work for `fetch::start`

Comment (molant) [reply packages/rule-https-only/src/rule.ts]: ok, in any case we should be changing to the `fetch::end` so does this matter?

File: packages/rule-https-only/src/rule.ts (hunk 22 above)
Comment (molant) [reply packages/rule-https-only/src/rule.ts]: Yes, that's what I think. And check the `hops` property of `response`

Comment (sarvaje) [reply packages/rule-https-only/src/rule.ts]: ok

File: packages/rule-https-only/package.json (hunk 21 above)
Comment (sarvaje) [reply packages/rule-https-only/package.json]: The `package.json` in main doesn't include `@types/debug`

File: packages/rule-https-only/src/rule.ts (hunk 22 above)
Comment (molant) [reply packages/rule-https-only/src/rule.ts]: Only the target

**Comment by sarvaje** [comment]
@molant @alrra this should be ready to review

File: packages/rule-https-only/README.md [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+
+`https-only` checks if your site is using HTTPS and warns against having
+mixed content.
+
+## Why is this important?
+
+HTTPS is important to guarantee content integrity. Even when your site
+doesn't have sensitive information, an attacker can change the content
+or inject malicious scripts (like a [crypto miner][crypto miner] to
+use your users CPU power).
```
Comment (alrra) [inline packages/rule-https-only/README.md]: We should also add a note that some browser features are not restricted to HTTPS only.

See also: https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts

File: packages/rule-https-only/src/rule.ts [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+
+                return;
+            }
+
+            if (anyHopHTTP(response)) {
+                await context.report(resource, null, `Shouldn't be redirected from HTTP`);
+            }
+        };
+
+        context.on('fetch::end::*', validate);
```
Comment (alrra) [inline packages/rule-https-only/src/rule.ts]: I don't think we are fetching `<audio>`, `<video>`, `<object>`..., so we are not validating a lot of things.

See: https://w3c.github.io/webappsec-mixed-content/

Comment (molant) [reply packages/rule-https-only/src/rule.ts]: I guess it will depend on the connector and if the case of media if it has `autoplay` enabled?

What I'm thinking on doing:
* Have a list with all the resources we've analyzed
* Subscribe to those elements events and check it there
* If we receive them via a `fetch::end`, use that information instead

**Comment by molant** [comment]
@alrra, @sarvaje I've updated the code and docs with the feedback. This should be ready for review. I've added checks for a bunch of stuff (`srcset`, `source`, `poster`, `data`). Hopefully I haven't missed anything (but adding them should be trivial).

@kshyju if you want to participate of code reviews feel free (and let me know so I can ping you).

Thanks!

File: packages/rule-https-only/src/rule.ts [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+        /** Returns an array with all the URLs in the given `srcset` attribute or an empty string if none. */
+        const parseSrcSet = (srcset: string): Array<string> => {
+            if (!srcset) {
+                return [];
+            }
+
+            const urls = srcset
+                .split(',')
+                .map((entry) => {
+                    return entry.split(' ')[0].trim();
```
Comment (sarvaje) [inline packages/rule-https-only/src/rule.ts]: you should `trim` entry, before the `split(' ')` to be sure you don't have any whitespace at the beginning of the string.

**Comment by sarvaje** [comment]
@molant LGTM too, but I can't approve because it is my PR hehehehe

**Comment by molant** [comment]
@alrra @kshyju want to take a look at this PR?

**Comment by molant** [comment]
@alrray I've updated the PR with your feedback. I'll merge this tomorrow unless there are any objections.
